Do not use mutable default arguments

marlamb opened the pull request:
Using mutable default arguments is a common Python problem, see e.g. https://docs.python-guide.org/writing/gotchas/@mutable-default-arguments

In this specific case the default argument even tries to setup some infrastructure settings at import time, which can potentially fail.

marlamb commented:
> @marlamb please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> 
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
>     1. **Definitions**.
>        **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>        that is delivered by You to Microsoft under this Agreement.
>        **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>        approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>        **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>        Project, including but not limited to communication on electronic mailing lists, source code control
>        systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>        discussing and improving that Project, but excluding communication that is conspicuously marked or
>        otherwise designated in writing by You as “Not a Submission.”
>        **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>        associated comments and documentation.
> 
>     2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>        Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>        described in Section 4 below), Submit to any Project.
> 
>     3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>        Should You wish to Submit materials that are not Your original work, You may Submit them separately
>        to the Project if You (a) retain all copyright and license information that was in the materials as You
>        received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>        containing materials of a third party:” followed by the names of the third party and any licenses or other
>        restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>        guidelines concerning Submissions.
> 
>     4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>        for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>        Submission is made in the course of Your work for an employer or Your employer has intellectual
>        property rights in Your Submission by contract or applicable law, You must secure permission from Your
>        employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>        Agreement will refer to You and the employer collectively. If You change employers in the future and
>        desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>        and secure permission from the new employer before Submitting those Submissions.
> 
>     5. **Licenses**.
> 
> 
>     * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>       indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>       Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>       the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>       parties.
> 
>     * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>       indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>       Your patent claims that are necessarily infringed by the Submission or the combination of the
>       Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>       import or otherwise dispose of the Submission alone or with the Project.
> 
>     * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>       No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>       granted by implication, exhaustion, estoppel or otherwise.
> 
> 
>     6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>        licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>        have disclosed under Section 3). You represent that You have secured permission from Your employer to
>        make the Submission in cases where Your Submission is made in the course of Your work for Your
>        employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>        law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>        have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>        You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>        REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>        EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>        PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>        NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 
>     7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>        You later become aware that would make Your representations in this Agreement inaccurate in any
>        respect.
> 
>     8. **Information about Submissions**. You agree that contributions to Projects and information about
>        contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>        information that You submit with Your Submission.
> 
>     9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>        the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>        Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>        exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>        defenses of lack of personal jurisdiction and forum non-conveniens.
> 
>     10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>         supersedes any and all prior agreements, understandings or communications, written or oral, between
>         the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree

baywet reviewed (changes requested):
Hi @marlamb,
Thanks for using kiota and for starting this work. 
First off, here is the link to the article https://docs.python-guide.org/writing/gotchas/?highlight=mutable%20default#mutable-default-arguments

This is probably an issue we have across all the kiota Python repositories. And maybe in multiple places here. 
We'd be more than happy to receive pull requests for other defects. 

For this one, to ensure a swift release, can you also bump the patch version in the project file and add an entry in the changelog please?

marlamb commented:
> Hi @marlamb, Thanks for using kiota and for starting this work. First off, here is the link to the article https://docs.python-guide.org/writing/gotchas/?highlight=mutable%20default#mutable-default-arguments
> 
> This is probably an issue we have across all the kiota Python repositories. And maybe in multiple places here. We'd be more than happy to receive pull requests for other defects.
> 
> For this one, to ensure a swift release, can you also bump the patch version in the project file and add an entry in the changelog please?

Yes, it seems to be a common problem. I just opened two other pull requests (outside kiota, but within the microsoft org)
* https://github.com/microsoftgraph/msgraph-sdk-python-core/pull/626
* https://github.com/microsoftgraph/msgraph-sdk-python/pull/800
That are just the ones that popped up in my current codebase. I assume there will be even more. Would be great, if the maintainers could take them as a start to cleanup also the other potential occurrences, I am lacking the time to fix more than what I encounter directly.

Sorry for mistyping the link, I fixed it in the PR description.

I added the changes as you request, please have a look if it meets your expectations.
I didn't know what date to put, so left a `TBD` for now. In case you would like something else, I can of course change it easily.

baywet commented on CHANGELOG.md:
```diff
@@ -5,6 +5,13 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [1.3.2] - TBD
```
```suggestion
## [1.3.2] - 2024-01-09
```

baywet commented on CHANGELOG.md:
```diff
@@ -5,6 +5,13 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [1.3.2] - 2024-01-09
```
```suggestion
## [1.3.2] - 2024-07-09
```

baywet reviewed (approved):
Thanks for making the changes!
